crates/sui-indexer-alt/src/lib.rs

```diff
@@ -0,0 +1,209 @@
+            .context("Failed to connect to database")?;
+
+        let (metrics, metrics_service) =
+            MetricsService::new(metrics_address, db.clone(), cancel.clone())?;
+        let ingestion_service =
+            IngestionService::new(ingestion_config, metrics.clone(), cancel.clone())?;
+
+        Ok(Self {
+            db,
+            metrics,
+            metrics_service,
+            ingestion_service,
+            committer_config,
+            first_checkpoint,
+            last_checkpoint,
+            enabled_pipelines: pipeline.into_iter().collect(),
+            cancel,
+            first_checkpoint_from_watermark: u64::MAX,
+            handles: vec![],
+        })
+    }
+
+    /// Adds a new pipeline to this indexer and starts it up. Although their tasks have started,
+    /// they will be idle until the ingestion service starts, and serves it checkpoint data.
+    pub async fn pipeline<H: Handler + 'static>(&mut self) -> Result<()> {
+        if !self.enabled_pipelines.is_empty() && !self.enabled_pipelines.contains(H::NAME) {
+            info!("Skipping pipeline {}", H::NAME);
+            return Ok(());
+        }
+
+        let mut conn = self.db.connect().await.context("Failed DB connection")?;
+
+        let watermark = CommitterWatermark::get(&mut conn, H::NAME)
+            .await
+            .with_context(|| format!("Failed to get watermark for {}", H::NAME))?;
+
+        // TODO(amnn): Test this (depends on supporting migrations and tempdb).
+        self.first_checkpoint_from_watermark = watermark
+            .as_ref()
+            .map_or(0, |w| w.checkpoint_hi_inclusive as u64 + 1)
+            .min(self.first_checkpoint_from_watermark);
+
+        let (handler, committer) = pipeline::<H>(
+            watermark,
+            self.committer_config.clone(),
+            self.db.clone(),
+            self.ingestion_service.subscribe(),
+            self.metrics.clone(),
+            self.cancel.clone(),
+        );
+
+        self.handles.push(handler);
+        self.handles.push(committer);
+
+        Ok(())
+    }
+
+    /// Start ingesting checkpoints. Ingestion either starts from the configured
+    /// `first_checkpoint`, or it is calculated based on the watermarks of all active pipelines.
+    /// Ingestion will stop after consuming the configured `last_checkpoint`, if one is provided,
+    /// or will continue until it tracks the tip of the network.
+    pub async fn run(mut self) -> Result<JoinHandle<()>> {
+        let metrics_handle = self
+            .metrics_service
+            .run()
+            .await
+            .context("Failed to start metrics service")?;
+
+        // If an override has been provided, start ingestion from there, otherwise start ingestion
+        // from just after the lowest committer watermark across all enabled pipelines.
+        let first_checkpoint = self
+            .first_checkpoint
+            .unwrap_or(self.first_checkpoint_from_watermark);
```

Comment:
Do we (have we) actually do this rn? We do set the watermark on adding a new pipeline, but Indexer's first_checkpoint_from_watermark doesn't seem to be updated per the min committer watermark across all pipelines

Reply:
> Indexer's first_checkpoint_from_watermark doesn't seem to be updated per the min committer watermark across all pipelines

What about L146-149 above?